### PR TITLE
fix(plugin-threads): Fix comments UI & complimentary sidebar usability

### LINKED ISSUE #5159: [comments] no way to close right sidebar on mobile because button offscreen
repro:
- open composer on mobile
- toggle the right sidebar in top right
- you cannot close it, even with refresh

(for iOS Safari, the only way out is to delete all data for the dxos.org domain because right sidebar open state is stored in local storage)

### LINKED ISSUE #5237: [composer] Increment layout in preparation for Deck
This is the first chunk of several chunks of work being tackled to move Composer towards [the proposed design](https://www.figma.com/file/w9HsmcQattO4slk9fLWjnh/Operating-system-v1?type=design&node-id=42-479&mode=design&t=GyxcgUrV9mUnrLUA-0). The chunks being layout ( this ), attention ( #5698 ) & deck ( #5148 ).

- Move search to top of nav sidebar
  - CmdK as fallback if SearchPlugin is disabled
- Move self Avatar + Settings to bottom of nav sidebar
  - implement persistent notch
  - add ‘Beta’ instead of notifications trigger, keep tooltip, move version number into beta tooltip
- Remove persistent c11y sidebar trigger
- Refactor c11y sidebar to have dialog-style interactions
- Open c11y sidebar with parity:
  - If thread is present on navigate
  - If comments are present on navigate & focus new comment
- Close c11y sidebar if thread or comments are not present on navigate
- Move app footer to app settings tabpanel
  - Change ‘Done’ in settings dialog to cross close
  - Reduce max-bs of plugins tooltip (or increase padding around SettingsDialog is available)

### LINKED ISSUE #5323: [comments] show proper avatar in comments

### LINKED ISSUE #5340: [authoring] stories for comments (threads) and sidebar
- [ ] design increment for threads (focused on comments functionality but ensuring other threads features are maintained)
- [ ] stories for pure threads components
- [ ] use codemirror in messages for rendering
- [x] #5323 
- [ ] #5162
- [ ] ability to view/undo resolved threads
- [ ] #5161
- [ ] #5237

## Surfaces
- Sidebar showing chat channel thread (scrolling from bottom up)
- Sidebar showing multiple comment threads
- Main view showing comment thread
- Message card showing multi-part blocks: text, JSON, mosaic objects
- Message input with activity indicator
- Synchronized selection/scrolling of comment/editor selection ranges via intents

### PR DESCRIPTION
**Pull request by thure**
Resolves #5323. Resolves #5159. Resolves #5340.

https://github.com/dxos/dxos/assets/855039/a5e704a1-d72e-482b-b225-01447aa8d87c

This PR:
- [x] makes the necessary changes so comments look and behave as designed
  - [x] resolve outstanding technical debt
  - [x] implement threads as designed
  - [x] establish intermediate UI package for threads
  - [x] use `TextEditor` for the thread’s textbox
  - [x] move editor+comments story to ui thread package (since thread depends on editor, where that story is now)
  - [x] remove duplicate thread implementation from stories
  - [x] restore `onFocus` (probably to be renamed unless it is genuinely a focus event handler)
  - [x] restore activity indicator which is sometimes but not always distinct from disabling the thread’s textbox
  - [x] restore mosaic blocks in ThreadPlugin (probably as a Surface instead of just the title with a drag handle as it was previously)
  - [x] make all messages TextEditors (probably just in ThreadPlugin since this will relate to the Echo ontology)
  - [x] make pending state more visually obvious (opacity 50% perhaps, nontrivial to make TextEditor “look disabled”)
  - [x] add CommentThread titling components
- [x] improves c11y sidebar usability
  - [x] add persistent close button to sidebar, remove persistent toggle from topbar
- [ ] new componentry has parity
  - [x] check that this works in Composer
  - [ ] if there are any e2e tests for these features, ensure they are still up

This PR also:
- [x] Adds `dtLocale` to `TranslationsContext` and to the return value of `useTranslation` for use with `date-fns` functions
- [x] Upgrades Tailwind and `tailwind-merge` to [support `subgrid` classnames](https://tailwindcss.com/blog/tailwindcss-v3-4), which I anticipate we will use increasingly often, also support for some other quite powerful features that could be useful.

### CODE REVIEW

File: packages/ui/react-ui-thread/src/types.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+export type MessageMetadata = {
+  id: string;
+  authorId: string;
+  authorName?: string;
+  authorImgSrc?: string;
+  authorStatus?: AvatarStatus;
+};
+
+export type MessageEntityBlock = {
```
Comment (richburdon) [inline packages/ui/react-ui-thread/src/types.ts:15]: we have to use the existing reactive types here
i think it's likely premature to factor threads out into a separate package until we complete the schema work.

File: packages/ui/react-ui-thread/src/Message/Message.tsx [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+};
+
+export type MessageTextboxProps = {
+  disposition?: 'comment' | 'message';
+  onSend?: (text: string) => boolean | void;
+  pending?: boolean;
+} & Pick<MessageMetadata, 'id' | 'authorId' | 'authorImgSrc' | 'authorName'> &
+  Pick<TextEditorProps, 'model'>;
+
+export const MessageTextbox = (props: MessageTextboxProps) => {
```
[the remark this answers is not in the document]
Comment (thure) [reply packages/ui/react-ui-thread/src/Message/Message.tsx:130]: This PR is not ready for review, I will implement these affordances soon.

File: packages/ui/react-ui-thread/src/Message/Message.tsx [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+  const { t } = useTranslation(translationKey);
+  return (
+    <MessageMeta {...props} authorStatus='active' continues={false}>
+      <TextEditor
+        model={props.model}
+        placeholder={t(props.disposition === 'comment' ? 'comment placeholder' : 'message placeholder')}
```
[the remark this answers is not in the document]
Comment (thure) [reply packages/ui/react-ui-thread/src/Message/Message.tsx:136]: Sure, that’s fair enough.

File: packages/ui/react-ui-thread/src/types.ts (hunk 1 above)
Comment (thure) [reply packages/ui/react-ui-thread/src/types.ts]: I’ve followed Table’s precedent here, making this into a generic that consumers can enforce their type-safety with.

File: packages/apps/plugins/plugin-thread/src/components/MessageContainer.tsx [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+import { Mosaic, type MosaicTileComponent } from '@dxos/react-ui-mosaic';
+import { hoverableControlItem, hoverableControls, hoverableFocusedWithinControls, mx } from '@dxos/react-ui-theme';
+import { Message, type MessageBlockProps, type MessageProps } from '@dxos/react-ui-thread';
+
+import { useMessageMetadata } from '../hooks';
+import { THREAD_ITEM } from '../meta';
+
+type Block = MessageType.Block;
+
+const safeParseJson = (data: string) => {
```
Comment (wittjosiah) [inline packages/apps/plugins/plugin-thread/src/components/MessageContainer.tsx:24]: TODO to make util (this isn't the first time https://github.com/dxos/dxos/blob/main/packages/apps/composer-app/src/native.ts#L88)

File: packages/apps/plugins/plugin-thread/src/components/MessageContainer.tsx [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+      </div>
+    );
+  },
+);
+
+const TextboxBlock = ({
+  defaultContent,
+  onBlockDelete,
+}: { defaultContent?: string } & Pick<MessageBlockProps<Block>, 'onBlockDelete'>) => {
+  const [textObject] = useState(new TextObject(defaultContent));
```
Comment (wittjosiah) [inline packages/apps/plugins/plugin-thread/src/components/MessageContainer.tsx:71]: I'm a little confused by this. I think the prop should be a text object itself? These will come straight from the db as text objects in comments that we want to be live editable and going to string and back to text object seems likely to break something.

File: packages/apps/plugins/plugin-thread/src/components/ThreadContainer.tsx [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+import { MessageContainer } from './MessageContainer';
+import { useStatus, useMessageMetadata } from '../hooks';
+
+export type ThreadContainerProps = {
+  space: Space;
+  thread: ThreadType;
+  currentRelatedId?: string;
+  onAttend?: () => void;
+  autoFocusTextBox?: boolean;
```
Comment (wittjosiah) [inline packages/apps/plugins/plugin-thread/src/components/ThreadContainer.tsx:31]: jsdoc on these to explain their usage would be helpful

File: packages/apps/plugins/plugin-thread/src/components/ThreadContainer.tsx [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+      message.blocks.splice(index, 1);
+      if (message.blocks.length === 0) {
+        thread.messages.splice(messageIndex, 1);
+      }
+    }
+  };
+
+  const textboxMetadata = useMessageMetadata(thread.id, identity);
+
+  console.log('[thread current]', current, thread.id);
```
Comment (wittjosiah) [inline packages/apps/plugins/plugin-thread/src/components/ThreadContainer.tsx:96]: 👀

File: packages/apps/plugins/plugin-thread/src/components/ThreadsContainer.tsx [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+import { ThreadContainer, type ThreadContainerProps } from './ThreadContainer';
+
+export type ThreadsContainerProps = Omit<
+  ThreadContainerProps,
+  'thread' | 'onAttend' | 'current' | 'autoFocusTextBox'
+> & {
+  threads: ThreadType[];
+  onThreadAttend?: (thread: ThreadType) => void;
+  currentId?: string;
+  autoFocusCurrentTextbox?: boolean;
```
Comment (wittjosiah) [inline packages/apps/plugins/plugin-thread/src/components/ThreadsContainer.tsx:17]: here as well

File: packages/apps/plugins/plugin-thread/src/components/index.ts [hunk 9]
```diff
@@ -8,8 +8,8 @@ import React from 'react';
 export const ThreadMain = React.lazy(() => import('./ThreadMain'));
 
 // TODO(wittjosiah): Suspense boundary for sidebar?
-export * from './Chat';
-export * from './Comments';
-export * from './MessageCard';
-export * from './MessageInput';
+export * from './ThreadMain';
```
Comment (wittjosiah) [inline packages/apps/plugins/plugin-thread/src/components/index.ts:11]: note that main is exported above lazily already

File: packages/apps/plugins/plugin-thread/src/translations.ts [hunk 10]
```diff
@@ -16,6 +16,7 @@ export default [
         'message placeholder': 'Enter message...',
         'comment placeholder': 'Enter comment...',
         'settings standalone label': 'Enable standalone thread creation',
+        'anonymous label': 'Anonymous',
```
Comment (wittjosiah) [inline packages/apps/plugins/plugin-thread/src/translations.ts:19]: Do we need this, vs generate name?

File: packages/ui/react-ui-editor/src/components/TextEditor/TextEditor.tsx [hunk 11]
```diff
@@ -259,10 +259,10 @@ export const MarkdownEditor = forwardRef<EditorView, TextEditorProps>(
 export const defaultSlots: TextEditorSlots = {
   root: {
     // TODO(burdon): Add focusRing by default/as property?
```
Comment (wittjosiah) [inline packages/ui/react-ui-editor/src/components/TextEditor/TextEditor.tsx:261]: remove todo?

File: packages/ui/react-ui-editor/src/hooks/useTextModel.ts [hunk 12]
```diff
@@ -30,6 +30,17 @@ export const useTextModel = (props: UseTextModelProps): EditorModel | undefined
   return model;
 };
 
+export const useInMemoryTextModel = ({
```
Comment (wittjosiah) [inline packages/ui/react-ui-editor/src/hooks/useTextModel.ts:33]: jsdoc for intended usage

File: packages/ui/react-ui-thread/src/Message/Message.tsx [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+
+import { command } from './command-extension';
+import { translationKey } from '../translations';
+import { type MessageEntity, type MessageEntityBlock, type MessageMetadata } from '../types';
+
+const avatarSize = 7;
+
+const messageCell = 'plb-1 min-is-0';
+
+const safeParseJson = (data: string) => {
```
Comment (wittjosiah) [inline packages/ui/react-ui-thread/src/Message/Message.tsx:27]: todo factor out

File: packages/apps/plugins/plugin-thread/src/translations.ts (hunk 10 above)
Comment (thure) [reply packages/apps/plugins/plugin-thread/src/translations.ts:19]: We do need this as a UI-level detail — ThreadPlugin will substitute with a generated name so this is never encountered.

File: packages/apps/plugins/plugin-thread/src/components/MessageContainer.tsx [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+  const model = useTextModel({ text: textObject });
+  const textboxWidth = onBlockDelete ? 'col-span-2' : 'col-span-3';
+  return (
+    <div
+      role='none'
+      className={mx('col-span-3 grid grid-cols-subgrid', hoverableControls, hoverableFocusedWithinControls)}
+    >
+      {model ? (
+        <TextEditor model={model} slots={{ root: { className: textboxWidth } }} />
```
[the remark this answers is not in the document]
Comment (thure) [reply packages/apps/plugins/plugin-thread/src/components/MessageContainer.tsx:75]: Resolved this during pairing.

**Pull request merged**